File: backend/src/services/microsoft-auth.js
import { ConfidentialClientApplication } from '@azure/msal-node';
import { Client } from '@microsoft/microsoft-graph-client';
import { TokenCredentialAuthenticationProvider } from '@microsoft/microsoft-graph-client/authProviders/azureTokenCredentials/index.js';
import { ClientSecretCredential } from '@azure/identity';
import dotenv from 'dotenv';

dotenv.config();

/**
 * Microsoft Authentication Service
 * Handles OAuth 2.0 authentication with Microsoft Identity Platform
 */
class MicrosoftAuthService {
  constructor() {
    this.clientId = process.env.MICROSOFT_CLIENT_ID;
    this.clientSecret = process.env.MICROSOFT_CLIENT_SECRET;
    this.tenantId = process.env.MICROSOFT_TENANT_ID;
    this.redirectUri = process.env.MICROSOFT_REDIRECT_URI;

    if (!this.clientId || !this.clientSecret || !this.tenantId) {
      console.warn('⚠️  Microsoft credentials not configured. Using mock data.');
      this.isConfigured = false;
      return;
    }

    this.isConfigured = true;

    // Initialize MSAL (Microsoft Authentication Library)
    this.msalConfig = {
      auth: {
        clientId: this.clientId,
        authority: `https://login.microsoftonline.com/${this.tenantId}`,
        clientSecret: this.clientSecret,
      },
    };

    this.msalClient = new ConfidentialClientApplication(this.msalConfig);

    // Initialize Azure Identity credential
    this.credential = new ClientSecretCredential(
      this.tenantId,
      this.clientId,
      this.clientSecret
    );

    console.log('✅ Microsoft Authentication Service initialized');
  }

  /**
   * Get access token using client credentials flow
   * This is for app-only access (no user login required)
   */
  async getAccessToken() {
    if (!this.isConfigured) {
      throw new Error('Microsoft credentials not configured');
    }

    try {
      console.log('🔑 Requesting access token...');
      console.log('   Tenant ID:', this.tenantId);
      console.log('   Client ID:', this.clientId);
      
      const tokenRequest = {
        scopes: ['https://graph.microsoft.com/.default'],
      };

      const response = await this.msalClient.acquireTokenByClientCredential(tokenRequest);
      
      if (!response || !response.accessToken) {
        throw new Error('Failed to acquire access token');
      }

      console.log('✅ Access token acquired successfully');
      return response.accessToken;
    } catch (error) {
      console.error('❌ Error getting access token:');
      console.error('   Error message:', error.message);
      console.error('   Error code:', error.errorCode);
      console.error('   Status:', error.status);
      
      if (error.errorCode === 'invalid_client') {
        console.error('\n⚠️  INVALID CLIENT - Possible issues:');
        console.error('   1. Client Secret is incorrect or expired');
        console.error('   2. Client ID or Tenant ID is wrong');
        console.error('   3. Generate a new Client Secret in Azure Portal');
      }
      
      if (error.status === 401 || error.status === 403) {
        console.error('\n⚠️  AUTHENTICATION FAILED - Possible issues:');
        console.error('   1. API permissions not granted');
        console.error('   2. Admin consent not provided');
        console.error('   3. Wait 5-10 minutes after granting permissions');
      }
      
      throw error;
    }
  }

  /**
   * Create authenticated Microsoft Graph client
   */
  async getGraphClient() {
    if (!this.isConfigured) {
      throw new Error('Microsoft credentials not configured');
    }

    try {
      const accessToken = await this.getAccessToken();

      const client = Client.init({
        authProvider: (done) => {
          done(null, accessToken);
        },
      });

      return client;
    } catch (error) {
      console.error('Error creating Graph client:', error);
      throw error;
    }
  }

  /**
   * Check if Microsoft integration is properly configured
   */
  isReady() {
    return this.isConfigured;
  }
}

// Export singleton instance
export const authService = new MicrosoftAuthService();

File: python-backend/app/core/microsoft_auth.py
from typing import Optional, Dict, Any
import logging
import msal
from app.core.config import settings

logger = logging.getLogger(__name__)

class MicrosoftAuthService:
    def __init__(self):
        self.client_id = settings.MICROSOFT_CLIENT_ID
        self.client_secret = settings.MICROSOFT_CLIENT_SECRET
        self.tenant_id = settings.MICROSOFT_TENANT_ID
        
        if not self.client_id or not self.client_secret or not self.tenant_id:
            logger.warning("Microsoft credentials not fully configured.")
            self.is_configured = False
            return
            
        self.is_configured = True
        
        self.authority = f"https://login.microsoftonline.com/{self.tenant_id}"
        self.scopes = ["https://graph.microsoft.com/.default"]
        
        self.app = msal.ConfidentialClientApplication(
            self.client_id,
            authority=self.authority,
            client_credential=self.client_secret,
        )
        
        logger.info("Microsoft Authentication Service initialized")

    def get_access_token(self) -> str:
        if not self.is_configured:
            raise ValueError("Microsoft credentials not configured")
            
        # Check cache first (MSAL handles this internally securely)
        result = self.app.acquire_token_silent(self.scopes, account=None)

        if not result:
            logger.info("Requesting new access token from Microsoft...")
            result = self.app.acquire_token_for_client(scopes=self.scopes)

        if "access_token" in result:
            return result["access_token"]
        else:
            error_msg = result.get("error_description", "Unknown error")
            logger.error(f"Failed to acquire token: {error_msg}")
            raise Exception(f"Failed to acquire access token: {error_msg}")

    def is_ready(self) -> bool:
        return self.is_configured

# Singleton instance
microsoft_auth = MicrosoftAuthService()
